src/model/document/document.ts

```diff
@@ -5,12 +5,25 @@ import { Vector2 } from '../vectors'
 import { readonly, shallowReactive, shallowRef, ref, watch } from 'vue'
 import arrayUtils from '../array-utils'
 import { ScopeElement, ScopeElementType } from './elements/scope-element'
+import { ExpressionElement, ExpressionElementType } from './elements/expression-element'
+
+type JsonType = null | boolean | number | string | JsonType[] | Vector2 | { [prop: string]: JsonType }
+
+type SerializedDataType = {
+  expressionElements: JsonType[]
+  scopeElement: JsonType[]
+}
+
+// type DeSerializedDataType = {
+//   expressionElements: JsonType[]
+//   scopeElement: JsonType[]
+// }
 
 export type QuantumDocumentElementTypes<T extends readonly QuantumElementType[] = readonly QuantumElementType[]> = {
   [key in T[number]['typeName']]: T[number]
-} & { ['scope-element']: typeof ScopeElementType }
+} & { ['scope-element']: typeof ScopeElementType } & { ['expression-element']: typeof ExpressionElementType }
```

Comment:
This Typescript type is one of the more interesting/over-engineered ones in this project. On one hand, it's super general and would make creating documents with arbitrary elements work reasonably well. On the other hand, it's pretty complex as you've surely noticed.

The alternative would be simply spelling out the element types that a document can have.

What are your thoughts?

Reply:
Yes, this was very complex and I still don't understand it, I just found the pattern and continued it :). I don't know if the number of Element Types will grow too large so maybe it is ok.

Reply:
I think for now the easiest option is to keep this type.
If https://github.com/microsoft/TypeScript/issues/33892 gets implemented anytime soon, that'd make things a fair bit nicer. Then we could have code along the lines of this and could probably get rid of a fair amount of weird types.
```js
export class ExpressionElement extends QuantumElement {
  typeName: typeof ElementType = ElementType
  static serialize() {
    // put serialization logic here
  }

  static deserialize() {
   ...
}
```